File: evals/fixtures/go-mcp-server-single/repo/internal/tools/run_shell_command.go
// Package tools implements the MCP tool handlers exposed by karen-mcp-go:
// run_shell_command (diagnostic/remediation shell access) and read_config
// (bounded file reads under an allowed config root). Handlers receive
// caller-supplied arguments directly from whatever LLM client is connected
// over the MCP transport — that input is untrusted exactly like any other
// network-adjacent input and must be validated before it reaches a
// subprocess or a filesystem path.
package tools

import "os/exec"

// RunShellCommandArgs is the caller-supplied argument shape for the
// run_shell_command tool call, as delivered by whatever MCP client is
// connected over stdio.
type RunShellCommandArgs struct {
	Command string `json:"command"`
}

// Decoy: this warning is documentation only — it appears in a comment,
// never executes, and should not be flagged by a structural scanner.
// Bad example, do NOT do this: exec.Command("sh", "-c", "rm -rf "+userInput)

// RunShellCommand executes a caller-supplied shell command on the host and
// returns its combined output, for runbook automation triggered by a
// connected MCP client.
//
// Real issue (shell-injection): args.Command is spliced directly into a
// shell command string passed to `sh -c` with no allow-list, no escaping,
// and no args-array form — a connected LLM client (or anything spoofing
// tool-call input) can run arbitrary shell commands with this process's
// privileges.
func RunShellCommand(args RunShellCommandArgs) (string, error) {
	cmd := exec.Command("sh", "-c", args.Command)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", err
	}
	return string(out), nil
}

File: evals/fixtures/node-mcp-server-single/repo/src/tools/runShellCommand.ts
import { exec } from "node:child_process";

export const runShellCommandTool = {
  name: "run_shell_command",
  description: "Runs a shell command on the host and returns its stdout, for runbook automation.",
  inputSchema: {
    type: "object",
    properties: {
      command: { type: "string", description: "The shell command to run" },
    },
    required: ["command"],
  },
};

// Decoy: this warning is documentation only — it appears in a comment, never
// executes, and should not be flagged by a structural scanner.
// Bad example, do NOT do this: child_process.exec(`rm -rf ${userInput}`)

interface RunShellCommandArgs {
  command: string;
}

// Real issue (shell-injection): the tool handler passes the caller-supplied
// `command` string straight to child_process.exec with no allow-list, no
// escaping, and no execFile/args-array — a connected LLM client (or anything
// spoofing tool-call input) can run arbitrary shell commands on the host.
export function handleRunShellCommand(args: RunShellCommandArgs): Promise<{ content: Array<{ type: string; text: string }> }> {
  return new Promise((resolve, reject) => {
    exec(args.command, (err, stdout, stderr) => {
      if (err) {
        reject(err);
        return;
      }
      resolve({ content: [{ type: "text", text: stdout || stderr }] });
    });
  });
}
